src/transformers/models/maskformer/modeling_maskformer.py

```diff
@@ -0,0 +1,2492 @@
+        return down
+
+
+class MaskFormerFPNModel(nn.Module):
+    def __init__(self, in_features: int, lateral_widths: List[int], feature_size: int = 256):
+        """
+        Feature Pyramid Network, given an input tensor and a set of features map of different feature/spatial size, it
+        creates a list of features maps with the same feature size.
+
+        Args:
+            in_features (`int`):
+                The number of input features (channels).
+            lateral_widths (`List[int]`):
+                A list with the features (channels) size of each lateral connection.
+            feature_size (int, *optional*, defaults to 256):
+                The features (channels) of the resulting feature maps.
+        """
+        super().__init__()
+        self.stem = MaskFormerFPNConvLayer(in_features, feature_size)
+        self.layers = nn.Sequential(
+            *[MaskFormerFPNLayer(feature_size, lateral_width) for lateral_width in lateral_widths[::-1]]
+        )
+
+    def forward(self, features: List[Tensor]) -> List[Tensor]:
+        fpn_features = []
+        last_feature = features[-1]
+        other_features = features[:-1]
+        output = self.stem(last_feature)
+        for layer, left in zip(self.layers, other_features[::-1]):
+            output = layer(output, left)
+            fpn_features.append(output)
+        return fpn_features
+
+
+class MaskFormerPixelDecoder(nn.Module):
+    def __init__(self, *args, feature_size: int = 256, mask_feature_size: int = 256, **kwargs):
+        """
+        Pixel Decoder Module proposed in [Per-Pixel Classification is Not All You Need for Semantic
+        Segmentation](https://arxiv.org/abs/2107.06278). It first runs the backbone's feature into a Feature Pyramid
+        Network creating a list of features maps. Then, it projects the last one to the correct `mask_size`.
+
+        Args:
+            feature_size (`int`, *optional*, defaults to 256):
+                The feature size (channel dimension) of the FPN feature maps.
+            mask_feature_size (`int`, *optional*, defaults to 256):
+                The features (channels) of the target masks size \\C_{\epsilon}\\ in the paper.
+        """
+        super().__init__()
+        self.fpn = MaskFormerFPNModel(*args, feature_size=feature_size, **kwargs)
+        self.mask_projection = nn.Conv2d(feature_size, mask_feature_size, kernel_size=3, padding=1)
+
+    def forward(self, features: List[Tensor], output_hidden_states: bool = False) -> MaskFormerPixelDecoderOutput:
+        fpn_features: List[Tensor] = self.fpn(features)
+        # we use the last feature map
+        last_feature_projected = self.mask_projection(fpn_features[-1])
+        return MaskFormerPixelDecoderOutput(
+            last_hidden_state=last_feature_projected, hidden_states=tuple(fpn_features) if output_hidden_states else ()
+        )
+
+
+# copied and adapted from original implementation, also practically equal to DetrSinePositionEmbedding
```

Comment:
Can a "copied from" statement be added here?

Reply:
no because I used `torch.pi` and not `math.pi`

Reply:
Sure, and is `math.pi` not equivalent to `torch.pi`?